### ISSUE TITLE
Reuse component if the new path resolves to the same route

### ISSUE THREAD

**Issue by buremba** [nuxt/nuxt.js#5062]
### What problem does this feature solve?

I'm aware of the watchQuery parameter but it only works for the query parameters. Consider the following case:

`/report` and `/report/10` resolves to the same page component and when the user is in the `/report` page and saves the report, we redirect the user to the `/report/10` but prevent Nuxt to reload the page component as we have all the data and the state is updated.

I tried to use `watchQuery` but it didn't work so I had to try native `history.pushState` but it seems to cause trouble in navigation. (user triggered back/forward buttons).

### What does the proposed changes look like?

If `/report` and `/report/10` resolves to the same route, when I call `$route.push`, I should be able to reuse the component so the page reload should not be triggered when an optional flag is enabled.

<div align="right"><sub><em>This feature request is available on <a href="https://cmty.app/nuxt">Nuxt</a> community (<a href="https://cmty.app/nuxt/nuxt.js/issues/c8686">#c8686</a>)</em></sub></div>

**Comment by manniL** [comment]
Duplicate of https://github.com/nuxt/nuxt.js/issues/4663

**Comment by buremba** [comment]
@manniL Unfortunately, the previous issue doesn't provide any reliable solution for this feature.

The first problem with the `key` approach is that on the latest version of Nuxt, it calls `data()` so reset the state even though we have `key` attribute which is a static value.

Also setting the value of `key` automatically disable the navigation and every `$route.push` that resolves to the same components ends up being impotent. It would be much better to have an API similar to `watchQuery`, the purpose of `key` seems to be different than this use-case. I would vote for the issue to be re-opened.

**Comment by manniL** [comment]
Makes sense. 👍

**Comment by buremba** [comment]
I realized that even though I redirect the user to another route, the page redirect doesn't work when `key` is defined.

**Comment by Atinux** [comment]
Hi @buremba 

In this particular scenario, you want to tell Nuxt to not load any `asyncData` from the matched page when you call `this.$router.push('/report/10')` right?

**Comment by buremba** [comment]
Hey @Atinux, correct but the issue seems to be more complex than the `asyncData` issue. Nuxt doesn't really load pages even though the path resolves to another route.

**Comment by webspilka** [comment]
Hi @buremba 
Can you look at my issue [https://github.com/nuxt/nuxt.js/issues/5257](url), Is it related to this discussion?

**Comment by buremba** [comment]
@webspilka I don't think so. This one is a feature request.

**Comment by johnRivs** [comment]
Side note: this would also somewhat help with #4132 because the transition (presumably) wouldn't trigger.

**Comment by buremba** [comment]
@johnRivs that seems to be the case. I guess that this issue has many side-effects.

**Comment by buremba** [comment]
Any updates on this issue?

**Comment by SHxKM** [comment]
I'm also awaiting updates on this issue, as I'm suffering the consequences of #4132 - API content appears -> transition occurs -> API content re-appears. In my case I can't use `beforeMount()` as suggested in that issue.

**Comment by sullivanpt** [comment]
An FYI, it's a hack, but I was able to work-around Nuxt `data()` refreshing my component data on route change by copying forward the data from the previous life-cycle as folows
```
export default {
  key: '_bar',
  data () {
    return {
       myVariable: this.myVariable || 'my default value'
    }
  }
}
```

**Comment by buremba** [comment]
We also use these tricks as a workaround but it caused other issues in our app. Is there anyone working on this issue? If not, the core maintainers let me know where to look, I can try to create a PR maybe?

**Comment by Atinux** [comment]
Hi @buremba 

You need to look into packages/vue-app/packages/vue-app/template/client.js
But his behavior is tricky since it bypasse middleware/validate. What about using the store to keep the last page data and avoid fetching again if the same data?

**Comment by buremba** [comment]
@Atinux yes, we again found a workaround but it's fragile and causes more issues as we upgrade Nuxt. :( I will look at it when I have time and try to send a PR, thanks!
